src/text-editor.coffee

```diff
@@ -2236,6 +2236,60 @@ class TextEditor extends Model
     return unless @getSoftTabs()
     @scanInBufferRange /\t/g, bufferRange, ({replace}) => replace(@getTabText())
 
+  # Essential: Set the indentation level for the given buffer row.
+  #
+  # Rebuilds the indentation based on the soft tabs and tab length settings of
+  # this editor in order to bring it to the given indentation level.
+  # Note that if soft tabs are enabled and the tab length is 2, a row with 4
+  # leading spaces would have an indentation level of 2.
+  #
+  # * `bufferRow` A {Number} indicating the buffer row.
+  # * `newLevel` A {Number} indicating the new indentation level.
+  setIndentLevelForRow: (bufferRow, newLevel) ->
+    @modifyIndentLevelForRow(bufferRow, newLevel)
+
+  # Essential: Adjust the indentation level for the given buffer row.
+  #
+  # Rebuilds the indentation based on the soft tabs and tab length settings of
+  # this editor in order to increase or decrease it by the requested number of
+  # levels.
+  # Note that if soft tabs are enabled and the tab length is 2, a row with 4
+  # leading spaces would have an indentation level of 2.
+  #
+  # * `bufferRow` A {Number} indicating the buffer row.
+  # * `adjustLevel` A {Number} indicating how many levels to increase by. A
+  #   negative value will remove levels.
+  adjustIndentLevelForRow: (bufferRow, adjustLevel) ->
+    @modifyIndentLevelForRow(bufferRow, undefined, adjustLevel)
+
+  # Helper for modifying indentation for a buffer row
+  modifyIndentLevelForRow: (bufferRow, newLevel=undefined, indentIncrease=0) ->
+    leadingWhitespace = @lineTextForBufferRow(bufferRow).match(/^\s*/)[0]
+    newIndentString = @modifyIndentWhitespace(leadingWhitespace, newLevel, indentIncrease)
+    @buffer.setTextInRange([[bufferRow, 0], [bufferRow, leadingWhitespace.length]], newIndentString)
```

Comment:
@izuzak I remember now why I did this :)

It was the keep the buffer change small, so the "patch" generated by @buffer.setTextInRange is kept small (not sure if would be a performance thing - though it means it benefits from future enhancements I think). Also, unless there's a way to set a line of text (I hadn't seen one) we'd still be calling @buffer.setTextInRange() anyway but with the whole line. So I felt may as well do the minimum change.